lib/Doctrine/DBAL/Driver/IBMDB2/DataSourceName.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Driver\IBMDB2;
+
+use function implode;
+use function sprintf;
+use function strpos;
+
+/**
+ * IBM DB2 DSN
+ */
+final class DataSourceName
+{
+    /** @var string */
+    private $string;
+
+    private function __construct(string $string)
+    {
+        $this->string = $string;
+    }
+
+    public function toString() : string
+    {
+        return $this->string;
+    }
+
+    /**
+     * Creates the object from an array representation
+     *
+     * @param array<string,mixed> $params
```

Comment:
Let's document the possible keys with an array shape?

Reply:
This method doesn't limit the keys that can be used.

Reply:
Array shapes do not limit the keys that can be used either

Reply:
So what will be the behavior if the key being passed is not mandatory or optional?

In this specific case, it's still a map of any string to any value that can be cast to a string. But instead of defining it as an array shape, I'd rather have an object/structure. If we don't do that soon, we may define an array shape later when bumping the Psalm level and it demands it. This way, we'll make sure that our type annotations are valid.

Reply:
> So what will be the behavior if the key being passed is not mandatory or optional?

Not the one I thought apparently… I could have sworn this was OK: https://psalm.dev/r/cd931bdda1

Reply:
No worries. I wasn't sure about how Psalm works either, so now we're on the same page :+1: